### PR TITLE
planner: refine range builder when meet `unsigned_int_col <cmp> -int_cnst`

### PR DESCRIPTION
**Pull request by b41sh**
### What problem does this PR solve? 

fix unexpected result of `unsigned_int_col >= -int_cnst` [#10450](https://github.com/pingcap/tidb/issues/10450)

### What is changed and how it works?

convert negative value to zero for unsigned column

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: util/ranger/ranger_test.go [hunk 1]
```diff
@@ -660,6 +660,13 @@ func (s *testRangerSuite) TestIndexRangeForUnsignedInt(c *C) {
 			filterConds: "[]",
 			resultStr:   `[(NULL,1) (2,9223372036854775810) (9223372036854775810,+inf]]`,
 		},
+		{
+			indexPos:    0,
+			exprStr:     `a >= -2`,
```
Comment (XuHuaiyu) [inline util/ranger/ranger_test.go]: If `a == -2`, we'll get `[eq(test.t.a), 0]`? 
Does it seem to be wrong?

Comment (b41sh) [reply util/ranger/ranger_test.go]: no
If a == -2, we'll get [eq(test.t.a), -2]
convertPoint function only affect ranges

File: util/ranger/ranger_test.go [hunk 2]
```diff
@@ -660,6 +660,20 @@ func (s *testRangerSuite) TestIndexRangeForUnsignedInt(c *C) {
...
+			accessConds: "[ge(test.t.a, -2)]",
+			filterConds: "[]",
+			resultStr:   `[[0,+inf]]`,
+		},
+		{
+			indexPos:    0,
+			exprStr:     `a = -2`,
+			accessConds: "[eq(test.t.a, -2)]",
+			filterConds: "[]",
+			resultStr:   `[[0,0]]`,
```
Comment (winoros) [inline util/ranger/ranger_test.go]: This case we should return empty range in fact?

Comment (XuHuaiyu) [reply util/ranger/ranger_test.go]: So is https://github.com/pingcap/tidb/issues/10447.

Comment (b41sh) [reply util/ranger/ranger_test.go]: yes
[[0,0]] is incorrect
I will modify my code

**Comment by b41sh** [comment]
@XuHuaiyu @winoros PTAL

**Comment by shenli** [comment]
@b41sh Could you add some comments and test cases?

File: util/ranger/points.go [hunk 3]
```diff
@@ -250,6 +250,36 @@ func (r *builder) buildFormBinOp(expr *expression.ScalarFunction) []point {
...
+		needFix := false
+		zeroValue := new(types.Datum)
+		if mysql.IsIntegerType(ft.Tp) && value.Kind() == types.KindInt64 {
+			needFix = true
+			zeroValue.SetInt64(int64(0))
+		} else if (ft.Tp == mysql.TypeDecimal || ft.Tp == mysql.TypeNewDecimal) &&
+			value.Kind() == types.KindMysqlDecimal {
+			needFix = true
+			zeroValue.SetMysqlDecimal(new(types.MyDecimal))
+		}
```
Comment (zz-jason) [inline util/ranger/points.go]: should we also consider float or double values?

Comment (b41sh) [reply util/ranger/points.go]: we don't need consider float and decimal, only unsigned integer will convert a negative number to positive number in function `func (d *Datum) convertToUint(sc *stmtctx.StatementContext, target *FieldType) (Datum, error)`

**Comment by XuHuaiyu** [comment]
It'll be better to add the integration tests mentioned in https://github.com/pingcap/tidb/issues/10450.

File: util/ranger/points.go [hunk 4]
```diff
@@ -250,6 +250,28 @@ func (r *builder) buildFormBinOp(expr *expression.ScalarFunction) []point {
 		return nil
 	}
 
+	// if the colunm is unsigned integer and the value
+	// is less than zero, we only need compare with zero.
+	if mysql.HasUnsignedFlag(ft.Flag) {
```
Comment (XuHuaiyu) [inline util/ranger/points.go]: Can we extract the part as a function like `handleUnsignedIntCol`?

**Comment by XuHuaiyu** [comment]
PTAL @winoros

File: util/ranger/points.go [hunk 5]
```diff
@@ -250,6 +250,28 @@ func (r *builder) buildFormBinOp(expr *expression.ScalarFunction) []point {
...
+		if mysql.IsIntegerType(ft.Tp) && value.Kind() == types.KindInt64 {
+			zeroValue := new(types.Datum)
+			zeroValue.SetUint64(0)
+			cmp, err := value.CompareDatum(r.sc, zeroValue)
```
Comment (winoros) [inline util/ranger/points.go]: Why not compare `value.GetInt64()` with 0 directly?

File: util/ranger/points.go (hunk 4 above)
Comment (b41sh) [reply util/ranger/points.go]: ok

File: util/ranger/points.go (hunk 5 above)
Comment (b41sh) [reply util/ranger/points.go]: ok

**Comment by b41sh** [comment]
I have extract the code as a function `handleUnsignedIntCol` PTAL @XuHuaiyu @winoros

File: util/ranger/points.go [hunk 6]
```diff
@@ -361,6 +344,28 @@ func HandlePadCharToFullLength(sc *stmtctx.StatementContext, ft *types.FieldType
...
+		return val, op, nil
+	}
+
+	// if the operation is GT, GE or NE, the range is [0, +inf]
+	// otherwise the value can not match any (key, value) pair in tikv storage.
+	if op == ast.GT || op == ast.GE || op == ast.NE {
+		op = ast.GE
+		val.SetUint64(0)
+	} else {
+		return val, op, kv.ErrNotExist
```
Comment (XuHuaiyu) [inline util/ranger/points.go]: Why do we return kv.ErrNotExist here?
Can we return a bool flag like `isValidRange` to indicate whether this range is valid or invalid?

Comment (b41sh) [reply util/ranger/points.go]: modified

File: util/ranger/points.go [hunk 7]
```diff
@@ -339,6 +344,28 @@ func HandlePadCharToFullLength(sc *stmtctx.StatementContext, ft *types.FieldType
...
+func handleUnsignedIntCol(ft *types.FieldType, val types.Datum, op string) (types.Datum, string, bool) {
+	isUnsigned := mysql.HasUnsignedFlag(ft.Flag)
+	isIntegerType := mysql.IsIntegerType(ft.Tp)
+	isNegativeInteger := (val.Kind() == types.KindInt64 && val.GetInt64() < 0)
+
+	if !isUnsigned || !isIntegerType || !isNegativeInteger {
+		return val, op, true
+	}
+
+	// if the operation is GT, GE or NE, the range is [0, +inf]
```
Comment (winoros) [inline util/ranger/points.go]:
```suggestion
	// If the operation is GT, GE or NE, the range should be [0, +inf].
```

File: util/ranger/points.go [hunk 8]
```diff
@@ -339,6 +344,28 @@ func HandlePadCharToFullLength(sc *stmtctx.StatementContext, ft *types.FieldType
...
+	// otherwise the value is not in valid range
```
Comment (winoros) [inline util/ranger/points.go]:
```suggestion
	// Otherwise the value is out of valid range.
```

File: util/ranger/points.go [hunk 9]
```diff
@@ -339,6 +344,28 @@ func HandlePadCharToFullLength(sc *stmtctx.StatementContext, ft *types.FieldType
 	}
 }
 
+// handleUnsignedIntCol handles negative integer for unsigned integer colunm
```
Comment (winoros) [inline util/ranger/points.go]: Add some more comments for three returned values.

Comment (b41sh) [reply util/ranger/points.go]: done

File: util/ranger/points.go [hunk 10]
```diff
@@ -339,6 +344,29 @@ func HandlePadCharToFullLength(sc *stmtctx.StatementContext, ft *types.FieldType
...
+
+	if !isUnsigned || !isIntegerType || !isNegativeInteger {
+		return val, op, true
+	}
+
+	// If the operation is GT, GE or NE, the range should be [0, +inf].
+	// Otherwise the value is out of valid range.
+	if op == ast.GT || op == ast.GE || op == ast.NE {
+		op = ast.GE
+		val.SetUint64(0)
```
Comment (XuHuaiyu) [inline util/ranger/points.go]: `return val, op, true` here and the else can be removed.

File: util/ranger/points.go [hunk 11]
```diff
@@ -250,6 +250,11 @@ func (r *builder) buildFormBinOp(expr *expression.ScalarFunction) []point {
 		return nil
 	}
 
+	value, op, isValidRange := handleUnsignedIntCol(ft, value, op)
+	if isValidRange == false {
```
Comment (XuHuaiyu) [inline util/ranger/points.go]: if !isValidRange

File: util/ranger/points.go (hunk 9 above)
Comment (winoros) [inline util/ranger/points.go]:
```suggestion
// handleUnsignedIntCol handles the case when unsigned column meets negative integer value.
```

File: util/ranger/points.go [hunk 12]
```diff
@@ -339,6 +344,28 @@ func HandlePadCharToFullLength(sc *stmtctx.StatementContext, ft *types.FieldType
 	}
 }
 
+// handleUnsignedIntCol handles negative integer for unsigned integer colunm
+// returns fixed value, op and isValidRange.
```
Comment (winoros) [inline util/ranger/points.go]:
```suggestion
// The three returned values are: fixed constant value, fixed function name, and a boolean which indicates whether the range is valid or not.
```

Comment (b41sh) [reply util/ranger/points.go]: ok

**Pull request merged**